README.md
# good-dag

Expandable append-only immutable directed acyclic graph using
linked lists.

There's no<sup>[1]</sup> way to get at the data if you throw away
the reference to the entry. Once every entry is garbage collected
the underlying data store will be as well, but until then, they
all stay in memory.

The references are one-way. Parents can't reference children, so
you can only walk _up_ the DAG towards the root, not down it to
the leaves.

This is a lowlevel fast implementation, suitable for storing
immutable data that is intended to stick around for the lifetime
of a task, and really not much else.

It's a tool for that one specific type of job.

_"Who's a good dag! YOU ARE!!"_

<small>[1]: _Technically_ that's not _exactly_ true, you could
poke around in `entry.data.entries[]`, but that's very "sticking
your fingers into the beast's mouth", and I do not recommend
it.</small>

## USAGE

```ts
// hybrid module, either works
import { GoodDAG } from 'good-dag'
// or
const { GoodDAG } = require('good-dag')

// default type is string, but you can store anything
const root = new GoodDAG<string>('root value')

// now the root value is set to 'root value'

// get is the root entry
assert.equal(root.value, 'root value')
assert.equal(root.parent(), undefined)

const child = root.append('child')
const otherChild = root.append('other child')

assert.equal(child.parent(), root)
assert.equal(otherChild.parent(), root)
```

## Class: `GoodDAG`

Represents a directed acyclic graph.

Note that child elements are actually instances of the `DAGEntry`
class, which `GoodDAG` implements. The two can be used
interchangeably, but this may confound `instanceof` checks, as
the two do not inherit from one another classically.

### `dag = new GoodDAG<T = string>(rootValue: T, blockSize:number = 65536)`

Initial root value is set to `rootValue`.

Block size is set by `blockSize`, and must be less than or equal
to `Math.pow(2, 32)`.

If less than or equal to `256`, then the internal linked lists
will use a `Uint8Array`. If less than or equal to `Math.pow(2,
16)`, then it will use a `Uint16Array`. Otherwise, it will use a
`Uint32Array`.

Keep in mind that using a larger block size will not _always_
improve performance, and can dramatically increase memory
footprint! See **Performance, Memory, and `blockSize` Tuning**
below for guidance on setting this value appropriately.

### `readonly dag.data: DAGData`

A reference to the underlying data store where this DAGEntry
lives.

### `readonly index: number`

The index of this entry in the underlying data store.

### `dag.parent() => DAGEntry | undefined`

Get the parent entry for the dag entry. Will return `undefined`
for the root entry.

### `dag.value() => T`

Return the value set for this entry in the DAG.

### `dag.isRoot() => boolean`

Return true if this is the root entry in the DAG.

### `dag.root() => DAGEntry<T>`

Return the root entry in the DAG.

### `dag.size() => number`

Return count of total number of items stored in the DAG.

### `dag.append(value: T) => DAGEntry<T>`

Adds a value into the DAG with `dag` as the parent, returning the
newly created entry.

Note that values are _not_ unique. Appending the same value
multiple times will result in the value appearing in the DAG
multiple times.

Also, the links are one-way, meaning that there is no way to get
to the children from the parent, only the other direction.

(It's very directed!)

## Class: `DAGEntry`

This is the class used for every entry _other than_ the root.

It is not intended to be instantiated directly, but it is
exported for the benefit of class-extension use cases,
`instanceof` type checking, and so on.

`DAGEntry` instances have all the same methods and properties as
`GoodDAG`.  The only difference is that they do not create an
underlying data store on creation, and instead use their
parent's.

Instantiate `DAGEntry` classes by calling `dag.append(data)`.

## Performance, Memory, and `blockSize` Tuning

Tuning the `blockSize` requires some consideration. The default
is set to a reasonable middle-ground, suitable for purposes where
you expect to have no more than around 50-100k entries in the
DAG.

The larger the `blockSize`, the more memory will be used for each
DAG entry. If using a `Uint8Array`, then `blockSize * 2` bytes
must be pre-allocated. If using `Uint16Array`, then `blockSize *
4`, and if `Uint32Array`, then `blockSize * 8`. This allocation
is usually quite fast, but if done repeatedly, the time will add
up, and a smaller block size will go faster.

However, when adding items to the DAG, once it goes beyond the
initial `blockSize` of items, a new data block must be allocated,
and parent lookups in that data block are ever so slightly
slower, because they may need to hop between blocks.

Some good rules of thumb: use the following block values based on
your best estimate of the most number of items will end up in the
DAG in your use case:

- 2000 items or less, or if you're at all memory constrained: set
  `blockSize` to `256`, so that each block only has to allocate
  512 bytes of memory. (1 byte word size, 2 words per block entry.)
- between 1000 and 10,000 items: set `blockSize` to `4096`. Each
  block will allocate 16kb. (2 byte word size, 2 words per
  entry.)
- between 10,000 and 50,000 items: set `blockSize` to `8192`.
  Each block will allocate 32kb.
- more than 50,000 items, especially if the DAG object isn't likely to
  stick around for a long time, or if you just don't want to blow
  more than 128kb per block: set `blockSize` to `65536`. Each
  block will allocate 128kb.  (Set it smaller than that if 128kb
  is more than you want to spend.)
- way more than 10m items (like billions), _and_ the DAG object
  is going to stick around for a long time (like a long-lived
  module-local variable, for example), so you won't hit the
  allocation operation very often, _and_ you have plenty of
  memory to burn, _and_ you really want to optimize lookup speed,
  _and_ you have a very confident limit on how many items you're
  ever going to need to store: set the limit to about a quarter
  of the total count of items you need to support. The total
  bytes allocated for each block will be this number times 8.
  For example, setting `blockSize` to 100,096 will require 782kb
  per block.

See
[`scripts/block-size-analysis.js`](scripts/block-size-analysis.js)
in this repo for a script to help understand these trade-offs.

## Things You Should Never Do With This Library

Don't pick a `blockSize` that is bigger than you need, especially
if you are creating brand new `GoodDAG` objects repeatedly.

Don't use this as a cache.  There's
[many](https://www.npmjs.com/package/mnemonist)
[other](http://npm.im/lru-cache)
[caches](http://npm.im/@isaacs/ttl-cache) that are actually
caches and quite good at what they do, so use one of those.

Don't use this as a key-value store, or get frustrated when it's
not that.

If your problem is DAG-shaped, this is a DAG-shaped tool.  It's a
good DAG, it's not a caravan.

## That is the joke.

[![dags, d'ya like dags?](./dags-do-you-like-dags.gif "dags, d'ya like dags?")](https://www.youtube.com/watch?v=zH64dlgyydM)


CHANGELOG.md
## v1.1

Export `DAGEntry` class

## 1.0

Initial release


test/index.ts
import t from 'tap'

import { DAGEntry, GoodDAG } from '../'

t.type(DAGEntry, 'function', 'DAGEntry class exported')

t.test('typechecks', t => {
  //@ts-expect-error
  new GoodDAG<string>(5, 5)
  //@ts-expect-error
  new GoodDAG()

  t.throws(() => new GoodDAG(true, -1))
  t.throws(() => new GoodDAG(true, 2.5))
  // if this fails it'll allocate 32gb of ram lol
  t.throws(() => new GoodDAG('hello', Math.pow(2, 32) + 1))
  t.throws(() => new GoodDAG(true, Math.pow(2, 32) - 1.5))
  t.end()
})

t.test('basic usage', t => {
  const root = new GoodDAG<number>(0, 5)
  t.type(root.data.parent, Uint8Array)
  t.equal(root.isRoot(), true)
  t.equal(root.parent(), undefined)
  t.equal(root.data.getParent(root.index), undefined)
  t.equal(root.root(), root)
  t.equal(root.value(), 0)
  const one = root.append(1)
  t.equal(one.root(), root)
  t.equal(one.isRoot(), false)
  const two = one.append(2)
  t.equal(two.root(), root)
  const tre = one.append(3)
  t.equal(tre.data, root.data, 'still on first block')
  const fur = two.append(4)
  t.equal(fur.value(), 4)
  const fiv = tre.append(5)
  t.equal(fiv.value(), 5)
  t.not(fiv.data, root.data, 'data overflowed to next block')
  t.equal(fiv.data, root.data.nextBlock, 'fiv on second block')
  const six = tre.append(6)
  t.equal(six.root(), root)
  t.equal(six.value(), 6)
  t.equal(root.size(), 7)
  t.equal(six.size(), 7)
  const otherSix = tre.append(6)
  t.equal(root.size(), 8)
  t.equal(six.size(), 8)
  t.equal(otherSix.value(), 6)
  t.equal(otherSix.parent(), tre)
  t.equal(six.parent(), tre)
  root.append(8)
  root.append(9)
  const ten = root.append(10)
  //@ts-ignore
  t.equal(ten.data, root.data.nextBlock.nextBlock, 'ten on third block')
  const lvn = six.append(11)
  t.equal(lvn.data.dags.length, 3, 'third block has refs to both others')
  t.equal(
    six.data.dags.length,
    2,
    'second block only refs first and itself'
  )
  t.end()
})

t.test('big dag', t => {
  const root = new GoodDAG('hello', Math.pow(2, 16) + 1)
  t.type(root.data.parent, Uint32Array)
  t.end()
})

t.test('make sure it does not run in circles chasing its tail', t => {
  // this is a really aggressive test.
  // Get into a situation where we have 256 external block
  // references in a single block.
  // Append the DAG with 255 entries all from the root,
  // filling the first block entirely.
  // Then, for each of those 256 entries, append another
  // 256 entries, completely filling a block for each one's children.
  // Then, go through each of the first 256 blocks, appending
  // one more entry, resulting in a final block with a completely
  // full parentDAG block.
  // Finally, go through every entry and ensure that we have not
  // corrupted any data, and that the total DAG size is 256*256+256,
  // proving that we are not limited by block count * word range in
  // the number of entries that can be stored.
  const root = new GoodDAG<string>('root', 256)
  // fill the first block entirely.
  for (let i = 0; i < 255; i++) {
    root.append(`gen-0 ${i}`)
  }
  t.equal(root.value(), 'root')
  t.equal(root.size(), 256)

  // add 256 entries for each
  for (const entry of root.data.entries) {
    for (let i = 0; i < 256; i++) {
      entry.append(`gen-1 ${i} parent=${entry.value()}`)
    }
  }
  t.equal(root.size(), 256 * 256 + 256)

  // now max out the parentDAG list in a final block.
  let data: DAGEntry<string>['data'] | undefined = root.data
  for (let i = 0; i < 256; i++) {
    const entry = data.entries[data.entries.length - 1]
    if (!data.nextBlock) {
      throw new Error('did not create 256 blocks!')
    } else {
      data = data.nextBlock
    }
    entry.append(`gen-2 ${i} parent=${entry.value()}`)
  }
  t.equal(root.value(), 'root')
  t.equal(root.size(), 256 * 256 + 256 + 256)

  data = root.data
  let block = 0
  do {
    // only snapshot the first and last blocks, the others are just
    // super repetitive and it blows up the fixture
    if (block === 0 || block === 257) {
      t.matchSnapshot(
        data.entries.map(e => e.value()),
        `block ${block}`
      )
    }
    block++
    data = data.nextBlock
  } while (data)
  t.equal(block, 258, 'have 258 blocks total')

  t.end()
})


src/index.ts
// a directed acyclic graph using a linked list
// when capacity is maxed, a new GoodDAG is created,
// and the link value is a reference to that other
// object and the index within it.

// exported class is a DAGEntry that creates its data storage
// and is the root entry.
export class GoodDAG<T = string> implements DAGEntry<T> {
  readonly data: DAGData<T>
  readonly index: number = 0

  constructor(rootValue: T, blockSize: number = max16) {
    this.data = new DAGData<T>(blockSize, rootValue)
    this.data.entries[0] = this
  }

  /**
   * get the parent entry to this one (returns undefined, this is the root)
   */
  parent() {
    return undefined
  }

  /**
   * get the value that this DAG entry stores
   */
  value(): T {
    return this.data.values[0]
  }

  /**
   * return true if this is the root entry (it is)
   */
  isRoot() {
    return true
  }

  /**
   * add a new entry into the DAG with the supplied value,
   * with this entry as its parent.
   */
  append(value: T) {
    return this.data.append(value, this.index, this.data)
  }

  /**
   * get the count of items stored in the DAG
   */
  size(): number {
    return this.data.size()
  }

  /**
   * get the root entry in the DAG (it's this object, this is the root)
   */
  root(): DAGEntry<T> {
    return this
  }
}

export class DAGEntry<T> {
  readonly data: DAGData<T>
  readonly index: number

  constructor(dag: DAGData<T>, index: number) {
    this.data = dag
    this.index = index
  }

  /**
   * get the value that this DAG entry stores
   */
  value(): T {
    return this.data.values[this.index]
  }

  /**
   * get the parent entry to this one
   */
  parent(): DAGEntry<T> | undefined {
    return this.data.getParent(this.index)
  }

  /**
   * return true if this is the root entry (it isn't)
   */
  isRoot(): boolean {
    return false
  }

  /**
   * add a new entry into the DAG with the supplied value,
   * with this entry as its parent.
   */
  append(value: T) {
    return this.data.append(value, this.index, this.data)
  }

  /**
   * get the count of items stored in the DAG
   */
  size(): number {
    return this.data.rootData().size()
  }

  /**
   * get the root entry in the DAG
   */
  root(): DAGEntry<T> {
    return this.data.rootData().entries[0]
  }
}

// double bitwise negation floors and casts to overflow 4-byte signed
const isLongInt = (i: number) => {
  const n: number = ~~i
  return n === i || (n <= 0 && max32 + n === i)
}
const isPosLongInt = (i: number) => i && isLongInt(i)

const max32 = Math.pow(2, 32)
const max16 = Math.pow(2, 16)
const max8 = Math.pow(2, 8)
const getUintArray = (max: number): UintArrayClass => {
  return max <= max8
    ? Uint8Array
    : max <= max16
    ? Uint16Array
    : Uint32Array
}
type UintArray = Uint8Array | Uint16Array | Uint32Array
type UintArrayClass = { new (n: number): UintArray }

class DAGData<T> {
  // these can grow up to blockSize
  dags: DAGData<T>[]
  entries: DAGEntry<T>[]
  values: T[]

  // these are initialized to blockSize at the start
  parent: UintArray
  parentDAG: UintArray
  blockSize: number
  // we always have a root entry, so nextFree starts at 1
  nextFree: number = 1
  nextBlock?: DAGData<T>

  constructor(
    blockSize: number,
    rootValue: T,
    parentDAG?: DAGData<T>,
    parentIndex: number = 0
  ) {
    if (!isPosLongInt(blockSize)) {
      throw new TypeError(
        'invalid block size, must be number less than 2**32'
      )
    }
    this.blockSize = blockSize
    const Arr: UintArrayClass = getUintArray(blockSize)

    // the zero-index value is rootValue, in parent or locally
    this.dags = [this]
    this.parent = new Arr(blockSize)
    this.parentDAG = new Arr(blockSize)

    this.values = [rootValue]

    const rootEntry = new DAGEntry<T>(this, 0)
    this.entries = [rootEntry]

    if (parentDAG) {
      // first entry has remote parent
      this.parent[0] = parentIndex
      this.dags.push(parentDAG)
      this.parentDAG[0] = 1
    }
  }

  rootData(): DAGData<T> {
    let d: DAGData<T> = this
    while (d.parentDAG[0] !== 0) {
      d = d.dags[d.parentDAG[0]]
    }
    return d
  }

  size(): number {
    return this.nextFree + (this.nextBlock?.size() || 0)
  }

  append(
    value: T,
    parentIndex: number,
    dag: DAGData<T> = this
  ): DAGEntry<T> {
    if (this.nextFree >= this.blockSize) {
      // ran out of space, add to next block
      if (!this.nextBlock) {
        // create next block with parent entry as root entry
        this.nextBlock = new DAGData<T>(
          this.blockSize,
          value,
          dag,
          parentIndex
        )
        return this.nextBlock.entries[0]
      } else {
        return this.nextBlock.append(value, parentIndex, dag)
      }
    } else {
      const index = this.nextFree++
      const entry = new DAGEntry<T>(this, index)
      this.entries[index] = entry
      this.parent[index] = parentIndex
      this.values[index] = value
      if (dag !== this) {
        // XXX: maybe faster to just always push and have dupes?
        // Benchmark this.
        const i = this.dags.indexOf(dag)
        const dagIndex = i === -1 ? this.dags.push(dag) : i
        this.parentDAG[index] = dagIndex
      }
      return entry
    }
  }

  getParent(index: number): DAGEntry<T> | undefined {
    if (index === 0 && this.parentDAG[index] === 0) {
      // root entry
      return undefined
    }
    const parentDAG = this.dags[this.parentDAG[index]]
    const parentIndex = this.parent[index]
    return parentDAG.entries[parentIndex]
  }
}


scripts/block-size-analysis.js
// this is a simulation of the time required for various UintArray
// expandable linked-list storage schemes.
//
// When a linked list is implemented using such a data structure, where each
// entry is an index into the data block, the Uint size must be large enough to
// accommodate any indexes in the array. So, longer array, bigger indexes,
// larger pointer size.
//
// An expanding linked list data block implementation can accommodate an
// infinite number of entries in theory, but in practice this means that more
// data blocks must be used when there are more entries than the capacity of a
// single block.  This scenario simulates that behavior.
//
// In each scenario, a Uint8Array, Uint16Array, and Uint32Array are tested.  As
// many blocks are created as would be needed to store the "count" value
// provided.  Then, the "items" amount of items are added to the array.
//
// takeaways:
// - A smaller block size is better when the fill count is smaller.  This makes
//   sense, because there's less to allocate.  It's also less data usage
//   overall, because each entry is 1 byte instead of 2 or 4.  However, the
//   speed impact is pretty minimal, like around 5-10%, and it's an operation
//   that runs on the order of 1000 - 10000 operations PER MS, so saving 10%
//   of that is not *ever* going to be material.
//
// - A larger block size is better if the eventual total fill count is higher
//   than a smaller array can accommodate, as it saves having to do multiple
//   allocation operations.  Again, this makes sense, but the overall impact
//   doesn't start to be significant until the range of at least around 3-4x
//   the maximum block size.  Ie, the 8bit approach falls off around 1000
//   items, and the 16 bit approach falls off around 200k items.
//
// - If the arrays *don't* actually fill to expected capacity, then it's
//   MUCH faster to use a smaller block size, because so much less data is
//   allocated up front.  All this means is that while a bigger block size
//   is better if you *know* you're going to have a lot of data, it's also
//   important to make a reasonable guess the maximum count you're likely to
//   *actually* need to accommodate, and keep the block size no more than
//   that rough range, if possible.
//
//   For example, if you know you're going to have anywhere from 100 to 10k
//   items, it's best to define the maximum block size at around 5000, so
//   that you'll use 16 bit arrays, and only rarely need more than 1 or 2
//   blocks.
//
// - At the very extreme end, if you set your block size to Math.pow(2, 32),
//   which forces using Uint32Array, and then only add a single item to it,
//   you get less than 1 operation per ms, and use a whopping 17_179_869_184
//   (16GiB) of allocation for nothing.  Setting a block size of 256 and then
//   adding a single entry is around 5000 operations per ms, by comparison,
//   and allocates only 256 bytes.
//
// When in doubt, just set the block size to Math.pow(2, 16), which allocs
// super fast, and is reasonably fast up to a million entries or so, and then
// the blocks are only 64kb each.  If you are fairly confident that you'll
// have a much smaller number of items, like 1-1000, set the blockSize to
// 256, so each block only requires 256 bytes and allocates much faster.

const N = Math.pow(10, 2)

// fill up as many arrays as needed, and then read them all
const test = (Cls, word, size, count, items = count) => {
  const arrays = []
  const blockSize = Math.min(size, count)
  let bytes = 0
  for (let n = 0; n < items; n++) {
    const ns = n % size
    if (ns === 0) {
      bytes += blockSize
      arrays.push(new Cls(blockSize))
    }
    const a = arrays[arrays.length - 1]
    a[ns] = ns
  }
  return bytes * word
}

const testSec = (Cls, word, size, count, items = count) => {
  const p = performance.now()
  const end = p + 1000
  let runs = 0
  let time = 0
  let bytes
  for (let q = p; q < end; q = performance.now()) {
    bytes = test(Cls, word, size, count, items)
    runs++
    time = q - p
  }
  return [runs / time, bytes]
}

const bench = (count, items = count) => {
  // need it to be high enough to get some data
  console.log('\nallocation for:', count)
  console.log('filling items: ', items)
  console.log(
    ' 8:',
    testSec(Uint8Array, 1, Math.pow(2, 8), count, items),
    'fills/ms, bytes used'
  )
  console.log(
    '16:',
    testSec(Uint16Array, 2, Math.pow(2, 16), count, items),
    'fills/ms, bytes used'
  )
  console.log(
    '32:',
    testSec(Uint32Array, 4, Math.pow(2, 32), count, items),
    'fills/ms, bytes used'
  )
}

console.log(
  'results is [score, bytes] ' + '(bigger score, smaller bytes = better)'
)

bench(1)
bench(10)
bench(100)
bench(1_000)
bench(10_000)
bench(100_000)
bench(200_000)
bench(1_000_000)
bench(10_000_000)
bench(Math.pow(2, 32) + 1, 10_000_000)
bench(1, 1)
bench(10, 1)
bench(100, 1)
bench(1_000, 1)
bench(10_000, 1)
bench(100_000, 1)
bench(200_000, 1)
bench(1_000_000, 1)
bench(10_000_000, 1)
bench(Math.pow(2, 32) + 1, 1)
